src/EditorFeatures/Core/Implementation/MissingImportTracking/MissingImportTrackingTaggerProvider.MissingImportTrackingCodeAction.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Tags;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.MissingImportTracking
+{
+    internal partial class MissingImportTrackingTaggerProvider
+    {
+        private class MissingImportTrackingCodeAction : CodeAction
+        {
+            private readonly string _title;
+            private readonly Document _document;
+            private MissingImportTrackingCommitter _missingImportTrackingCommitter;
+
+            public MissingImportTrackingCodeAction(Document document, string title)
+            {
+                _document = document;
+                _title = title;
+            }
+
+            public override string Title => _title;
+
+            protected override Task<IEnumerable<CodeActionOperation>> ComputeOperationsAsync(CancellationToken cancellationToken)
+            {
+                // Invoked directly without previewing.
+                if (_missingImportTrackingCommitter == null)
+                {
+                    if (!TryInitializeRenameTrackingCommitter(cancellationToken))
+                    {
+                        return SpecializedTasks.EmptyEnumerable<CodeActionOperation>();
+                    }
+                }
+
+                var committerOperation = new MissingImportTrackingCommitterOperation(_missingImportTrackingCommitter);
+                return Task.FromResult(SpecializedCollections.SingletonEnumerable(committerOperation as CodeActionOperation));
+            }
+
+            protected override async Task<IEnumerable<CodeActionOperation>> ComputePreviewOperationsAsync(CancellationToken cancellationToken)
+            {
+                if (!TryInitializeRenameTrackingCommitter(cancellationToken))
+                {
+                    return Enumerable.Empty<CodeActionOperation>();
+                }
+
+                var project = await _missingImportTrackingCommitter.AddMissingImportsAsync(cancellationToken).ConfigureAwait(false);
+                return SpecializedCollections.SingletonEnumerable((CodeActionOperation)new ApplyChangesOperation(project.Solution));
+            }
+
+            private bool TryInitializeRenameTrackingCommitter(CancellationToken cancellationToken)
+            {
+                if (_document.TryGetText(out var text))
+                {
+                    var textBuffer = text.Container.GetTextBuffer();
+                    if (textBuffer.Properties.TryGetProperty(typeof(StateMachine), out StateMachine stateMachine))
```

Comment:
tagging @dpoeschl since it uses rename tracker.

Reply:
I don't believe it does.  It has it's own StateMachine type (which is just called the same thing).

Reply:
hmmm.. the method name is TryInitializeRenameTrackingCommitter.

Reply:
@heejaechang It doesn't use Rename Tracking directly, but I did borrow heavily from the rename tracking implementation.